pkg/skaffold/build/ko/build.go

```diff
@@ -0,0 +1,100 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package ko
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/google/ko/pkg/build"
+	"github.com/google/ko/pkg/publish"
+
+	// latestV1 "github.com/GoogleContainerTools/skaffold/pkg/skaffold/schema/latest/v1"
+	latestV1 "github.com/GoogleContainerTools/skaffold/pkg/skaffold/build/ko/schema"
+)
+
+// Build an artifact using ko
+func (b *Builder) Build(ctx context.Context, out io.Writer, a *latestV1.Artifact, ref string) (string, error) {
+	koBuilder, err := b.newKoBuilder(ctx, a)
+	if err != nil {
+		return "", fmt.Errorf("error creating ko builder: %w", err)
+	}
+
+	koPublisher, err := b.newKoPublisher(ref)
+	if err != nil {
+		return "", fmt.Errorf("error creating ko publisher: %w", err)
+	}
+	defer koPublisher.Close()
+
+	imageRef, err := b.buildAndPublish(ctx, a.ImageName, koBuilder, koPublisher)
+	if err != nil {
+		return "", fmt.Errorf("could not build and publish ko image %q: %w", a.ImageName, err)
+	}
+	fmt.Fprintln(out, imageRef.Name())
+
+	return b.getImageIdentifier(ctx, imageRef, ref)
+}
+
+// buildAndPublish the image using the ko builder and publisher.
+func (b *Builder) buildAndPublish(ctx context.Context, imageName string, koBuilder build.Interface, koPublisher publish.Interface) (name.Reference, error) {
+	importpath, err := getImportPath(imageName, koBuilder)
+	if err != nil {
+		return nil, fmt.Errorf("could not determine Go import path for ko image %q: %w", imageName, err)
+	}
+	imageMap, err := b.publishImages(ctx, []string{importpath}, koPublisher, koBuilder)
+	if err != nil {
+		return nil, fmt.Errorf("failed to publish ko image: %w", err)
+	}
+	imageRef, exists := imageMap[importpath]
+	if !exists {
+		return nil, fmt.Errorf("no built image found for Go import path %q build images: %+v", importpath, imageMap)
+	}
+	return imageRef, nil
+}
+
+// getImportPath determines the Go import path that ko should build.
+//
+// If the image name from the Skaffold config has the prefix `ko://`, then
+// treat the remainder of the string as the Go import path to build. This
+// matches current ko behavior for working with Kubernetes resource files, and
+// it will allow ko users to easily migrate to Skaffold without changing their
+// Kubernetes YAML files. See https://github.com/google/ko#yaml-changes.
+//
+// If the image name does _not_ start with `ko://`, determine the Go import
+// path of the image workspace directory.
```

Comment:
Would it make sense to allow the user to specify the import-path?  There are users with a monorepo that contains multiple images.  So something like:
```
build:
  artifacts:
    - image: my/image/one
      builder: ko
      importPath: cmd/image1
    - image: my/image/two
      builder: ko
      importPath: cmd/image2
```

(implicit `context: .`)

Reply:
Yes! The idea is that users can pass this in the image name using the `ko://` prefix like this:

```yaml
build:
  artifacts:
    - image: ko://example.com/my/image/one
      ko: {}
    - image: ko://example.com/my/image/two
      ko: {}
```

For the same `context: .`. WDYT?

I'll add some unit tests to show this.

Reply:
What should happen for non-ko:// image references?

Reply:
Ko needs an import path to build an image. So for non-`ko://` image references, `getImportPath()` determines the Go import path from the context/workspace directory. This works with arbitrary image names in `skaffold.yaml` - in other words, a non-`ko://` image name doesn't need to match the import path.

The former supports current ko users moving to Skaffold, as they likely already have Kubernetes manifests with ko image references in place of image names. The latter matches Skaffold behavior for other builders, and would be more familiar for existing Skaffold users who want to start using ko.

Reply:
Added some unit tests to show non-cwd context/workspace builds.

Reply:
I think some of these decisions for Ko made sense because it was a command-line builder.

I'm trying to figure out where `KO_DOCKER_REPO` and `--default-repo` fits in here?  Let's say we want to promote Ko as the default Go builder.  For example, we package Skaffold itself up as an image at `gcr.io/k8s-skaffold/skaffold`. How would I represent this?

I think currently you'd suggest it should look like:
```
  artifacts:
    - image: github.com/GoogleContainerTools/skaffold/cmd/skaffold
       ko: {}
```

and then I have to run `skaffold build --default-repo gcr.io/k8s-skaffold`?  And what if my import package doesn't match my desired image name?

```
build:
  artifacts:
    - image: gcr.io/k8s-skaffold/skaffold
       ko: 
         importPath: ./cmd/skaffold
```

Or are you thinking that I'd need to set up a `.ko.yaml`  with an `id` and `main`?

Reply:
The design proposal wasn't very clear on this point, so I've pushed another commit that describes how the image naming works.

The short version is that the ko builder follows the existing Skaffold image naming logic. This is intentional, so that existing Skaffold users can switch to the ko builder without changes to image names. So if someone runs `skaffold build --default-repo gcr.io/k8s-skaffold` with `image: skaffold` in their `skaffold.yaml` file using the ko builder, the resulting image name is `gcr.io/k8s-skaffold/skaffold`. This works regardless of what the Go import path is. And users don't have to specify the `KO_DOCKER_REPO` environment variable, they only specify the Skaffold default repo.

The ko builder also supports existing ko users by allowing `image` names in `skaffold.yaml` to be of the form `ko://import.path/to/main/function`, e.g., `ko://github.com/googlecloudplatform/skaffold/cmd/skaffold`. The `SanitizeImageName()` function turns that + the default repo into `gcr.io/k8s-skaffold/github.com/googlecontainertools/skaffold/cmd/skaffold`. This allows existing ko users (say for example the Tekton and Knative teams) to bring their existing manifests to Skaffold unchanged. ([Tekton example](https://github.com/tektoncd/pipeline/blob/v0.25.0/config/controller.yaml#L66), [Knative example](https://github.com/knative/serving/blob/main/config/core/deployments/activator.yaml#L41)).

Reply:
Using the `context` to specify the importPath is a bit unorthodox , but more importantly precludes users who use a different project layout such as with mixed languages.

The `context` is best thought of as the "location of execution", where the build is run from.  This is important in Go as it determines the location of the `go.mod` (and possibly the `.ko.yaml`?).  There are apparently some users who have layouts like:
```
    /go.mod            the normal go.mod
    /pkg/foo/go.mod    specifies a different go version and/or different dependencies
```
So running their build from `pkg/foo` produces different results.

I think the Jib builder is a better model, which provides a `project` field.  A similar `importPath` would seem to make sense to me.

Reply:
> Using the `context` to specify the importPath is a bit unorthodox

The proposal uses `context` to specify the directory of execution.

> precludes users who use a different project layout such as with mixed languages

Does it? For the multi-`go.mod` example, this configuration would work:

```yaml
build:
  artifacts:
  - image: root
    context: .
    ko: {}
  - image: pkg-foo
    context: ./pkg/foo
    ko: {}
```

The `root` image in that example would use the `go.mod` from the same directory as `skaffold.yaml`, while the `pkg-foo` image would use `./pkg/foo/go.mod`. The ko builder runs `go list` and `go build` in the context directory. This is the ko logic,  the ko builder passes the context directory as the `dir` argument to the `moduleInfo()` function: https://github.com/google/ko/blob/690533235ad4209fab6c88f44314760b102ce829/pkg/build/gobuild.go#L153 and https://github.com/google/ko/blob/690533235ad4209fab6c88f44314760b102ce829/pkg/build/gobuild.go#L365

Full import paths (e.g., `github.com/org/repo` aren't required in `skaffold.yaml`, and I don't think we need to ask users to provide them. Supporting `ko://`-prefixed import paths as image names in `skaffold.yaml` is specifically just for existing ko users who use ko to render/populate image names in Kubernetes manifests. For these users, this feature simplifies Skaffold adoption, where Skaffold takes over rendering responsibilities, with no change required to their existing k8s manifest templates.

`.ko.yaml` is only read from the current process working directory, or from `KO_CONFIG_PATH`. It supports overriding the default base image, and the base image per import path prefix: https://github.com/google/ko#configuration

Reply:
I see. Given this directory structure:

```
./go.mod
./skaffold.yaml
./pkg/foo/main.go    # <- main() is here
```

The Skaffold config would be:

```yaml
build:
  artifacts:
  - image: gcr.io/k8s-project/image
    context: ./pkg/foo
```

ko (and `go build`) executes in the `./pkg/foo directory, so `context` specifies the directory of execution.

Reply:
this feels a little bit like a paradigm rift between the way we use `context` for the other builders - if you tried to simply "plug and play" a different builder (e.g. GCB) using this specified `context`, I don't think it would work since skaffold generally treats this as the root of all the required source files for building the artifact.

that said, since `ko` is a little different I don't think we should keep blocking this implementation since we can always go back and change this later. I still think this would be clearer for users as an `importPath` field in the build configuration (especially users who don't know anything about `ko` except it's "a good way to build go apps"), but we can revisit as an enhancement later.

Reply:
Yes, you're right. It'll fail for non-local builders.

I'll dig into what the solution here could be and create another PR. Since users can specify import path via the `image` field (as `ko://import.path/here/pkg/foo`), adding an `importPath` field would result in two ways to provide the import path, and that could be confusing.

Google Cloud Buildpacks fail if there are two `main()` functions in different directories where both are not the context directory. There, the `GOOGLE_BUILDABLE` envvar must be set: https://github.com/GoogleCloudPlatform/buildpacks/blob/main/README.md#configuration